sdlib/d/gc/slab.d

```diff
@@ -87,30 +124,28 @@ private:
 		assert(size <= sg.size, "size exceeds alloc size!");
 
 		if (size == 0) {
-			disableMetadata();
+			if (_hasMetadata) {
+				e.disableMetadata(sg.index);
+				_hasMetadata = false;
+			}
+
 			return;
 		}
 
 		writePackedFreeSpace(freeSpacePtr, size);
-		enableMetadata();
-	}
-
-	void enableMetadata() {
-		assert(_allowsMetadata, "size class not supports slab metadata!");
-
 		if (!_hasMetadata) {
 			e.enableMetadata(sg.index);
 			_hasMetadata = true;
 		}
 	}
 
-	void disableMetadata() {
-		assert(_allowsMetadata, "size class not supports slab metadata!");
+	enum FinalizerBit = nativeToBigEndian!size_t(0x2);
 
-		if (_hasMetadata) {
-			e.disableMetadata(sg.index);
-			_hasMetadata = false;
-		}
+	@property
+	bool finalizerEnabled() {
+		// TODO: right now we fetch hasMetadata eagerly, but the FinalizerBit check
+		// is cheaper. However, its false positive rate is currently unknown.
```

Comment:
It's false positive rate can easily be inferred from reasoning I posted before...